[Frontend][Tensorflow] Add unique operator

ymwangg opened the pull request:
This PR adds the tensorflow `unique` operator as described in https://www.tensorflow.org/api_docs/python/tf/unique.

I'm not sure I follow the best practices. Comments and suggestions are welcome. @yongwww @kevinthesun @codeislife99

codeislife99 commented on python/tvm/relay/op/_transform.py:
```diff
@@ -885,3 +886,29 @@ def where_shape_func(attrs, inputs, _):
+    """
+    Shape func for unique operator.
+    """
+    return [_unique_shape_1(inputs[0]), _unique_shape_1(inputs[0]), _unique_shape_1(inputs[0]), _unique_shape_2(inputs[0])]
```
You can put this in the same shape func, You can return a tuple from the shape func and remove the list from this function.

ymwangg replied:
Sure, I'll make it simpler.

codeislife99 commented on python/tvm/relay/op/strategy/generic.py:
```diff
@@ -1395,3 +1395,28 @@ def cumsum_strategy(attrs, inputs, out_type, target):
+    """Wrap unique topi compute"""
+
+    def _compute_unique(attrs, inputs, _):
+        return topi_compute(inputs[0])
```
Any reason the shape func is defined but not being used here?

ymwangg replied:
Since `output[0]`,`output[1]`,`output[2]` have the same shape as `input[0]` and `output[3]` has shape `[1]`. We don't need the shape function in compute function.

codeislife99 commented on src/relay/op/algorithm/unique.cc:
```diff
@@ -0,0 +1,147 @@
+    return false;
+  }
+  std::vector<Type> fields;
+  fields.push_back(TensorType(data->shape, data->dtype));
```
Instead of having a full tensor and then following it with strided slice , how about we make it totally dynamic itself ?

ymwangg replied:
The main challenge is the shape function. This operator requires data structures like set to efficiently calculate the number of unique elements. I don't think that's supported in hybrid script. Even if we can do it and get the shape function. It looks weird that both shape function and the compute function do the same computations.

masahi commented:
Thanks, I was planning to work on unique next week, happy to collaborate.

I can add TIR unqiue impl both cpu and gpu later.

ymwangg commented:
> Thanks, I was planning to work on unique next week, happy to collaborate.
> 
> I can add TIR unqiue impl both cpu and gpu later. We can add relay boilarplate, temp impl in cpp, and tests in this PR.

That would be great!

masahi commented:
@ymwangg For a general op like `unique`, we should follow numpy API, rather than being too specific to TF. PyTorch unique should be supported by the same API. Framework specific details should go into the frontend.

Numpy and PyTorch supports `dim` argument to do unique on multidimensional input, but I don't think it's a good idea. So restricting to 1D, at least for the first implementation, sounds good to me.

We can implement `unique` via sorting and cumsum. If implemented this way, the same code works on both CPU and GPU. That's I'm planning to do, but if you feel brave, you can try that in this PR :slightly_smiling_face:

ymwangg commented:
@masahi Thanks for your comment.
Here's the algorithm that I came up with based on your suggestions.
```python
# topi
def unique(data, data_sorted, data_argsorted):
    output = [0] * len(data)
    count = [0] * len(data)
    first_occurrence = [len(data)] * len(data)
    inverse_indices = [0] * len(data)
    num_unique = 0
    # ir_builder
    for i in range(len(data)):
        if i == 0 or data_sorted[i] != data_sorted[i-1]:
            num_unique += 1
            output[num_unique-1] = data_sorted[i]
            first_occurrence[num_unique-1] = min(first_occurrence[num_unique-1], data_argsorted[i])
        count[num_unique-1] += 1
        inverse_indices[data_argsorted[i]] = num_unique - 1
    return output, count, first_occurrence, inverse_indices, num_unique

# tf front end
def tf_unique(data):
    output, count, first_occurrence, inverse_indices, num_unique = unique(data, np.sort(data), np.argsort(data))
    sorted_occurence_indices = np.argsort(first_occurrence) # relay.argsort
    new_output = [output[sorted_occurence_indices[i]] for i in range(num_unique)] # relay.take
    index_converter = np.argsort(sorted_occurence_indices) # relay.argsort
    new_inverse_indices = [index_converter[i] for i in inverse_indices] # relay.take
    return new_output, new_inverse_indices
```
It defines a topi function that is similar to `np.unique` but requires the sorted data and the argsort of the data. In the frontend, it needs to do argsort twice if we want to keep the unique elements in the order of their first occurrence.

Does this look good to you?

masahi commented:
I can be a lot simpler than that. Unique is basically sort + adjacent difference + exclusive scan. If you don't understand that statement, the following example should help. We have exclusive scan for CPU (`cumsum` op with `exclusive=True`), and GPU (see https://github.com/apache/tvm/pull/7303).

```
import numpy as np


def exclusive_scan(arr):
    return np.cumsum(arr) - arr


inp = np.random.randint(0, 10, size=(15,))
argsort_indices = np.argsort(inp)
sorted_inp = np.array([inp[i] for i in argsort_indices])
print("sorted input:", sorted_inp)
adj_diff = np.concatenate([[sorted_inp[0]],  np.diff(sorted_inp)])
print("adjacent difference:", adj_diff)
non_zero = adj_diff != 0
non_zero[0] = True  # the first element is always selected
ex_scan = exclusive_scan(non_zero)
print("exclusive scan:", ex_scan)

unique = np.zeros(inp.shape[0], dtype=np.int)
for i in range(inp.shape[0]):
    if non_zero[i] != 0:
        unique[ex_scan[i]] = inp[argsort_indices[i]]

print("num unique element:", ex_scan[-1] + 1)
print("unique:", unique)
```

Output:
```
sorted input: [0 0 0 4 5 5 6 6 6 6 6 7 8 8 9]
adjacent difference: [0 0 0 4 1 0 1 0 0 0 0 1 1 0 1]
exclusive scan: [0 1 1 1 2 3 3 4 4 4 4 4 5 6 6]
num unique element: 7
unique: [0 4 5 6 7 8 9 0 0 0 0 0 0 0 0]
```

codeislife99 commented:
Hey @masahi , can your example be extended to provide `counts` as well ? https://www.tensorflow.org/api_docs/python/tf/unique_with_counts

masahi commented:
Yes, it's possible but a bit complicated. PyTorch also has `return_counts` option https://pytorch.org/docs/stable/generated/torch.unique.html

I think for the first PR, not all options need to be implemented. We can follow up later.

I'm using PyTorch GPU impl as reference, see for example below on how they support count
https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/aten/src/ATen/native/cuda/Unique.cu#L60-L68

codeislife99 commented:
I see, I was interested in `counts` option and probably you might be as well , because `SparseSegmentSqrtN` or other variants `SparseLengthSum` / `EmbeddingBag` can be written as a combination of `take`, `expand`, `repeat`,`scatter_add` and `unique_count`. So I was interested if we could do this either in this PR or in a parallel PR(and later merge them)

ymwangg commented:
@masahi Thanks for the explanation and it is very helpful!
It looks like the main thing we need to do is to implement a `topi.adjacent_difference` op similar to `thrust::adjacent_difference`. And in the frontend, we do something like:
```python
sorted_data = relay.sort(data)
argsort_indices = relay.argsort(data)
adj_diff = relay.adjacent_difference(sorted_data, first_value=0, "not_equal")
ex_scan = relay.cumsum(adj_diff, exclusive=True)
inverse_indices = relay.scatter(data, argsort_indices, ex_scan)
unique = relay.scatter(data, ex_scan, sorted_data)
unique_sliced = relay.strided_slice(unique, [0], relay.take(ex_scan,[-1]), slice_mode="size")
return unique_sliced, inverse_indices
```
I saw PyTorch uses `thrust::unique` to get the `unique` array. I think we can use `relay.scatter` to do the same thing.

To support counting, it looks like we need to implement a `topi.unique_by_key` op similar to `thrust::unique_by_key`. I think maybe we can do it in a different PR and focus on `adjacent_difference` in this PR.

masahi commented:
For your first implementation, combination-based approach is ok. But `unique` is important enough that I think it deserves its own operator. Also implementation directly in ir builder will likely be faster. Supporting other options will also be easier if we write in ir builder.

So use ir builder if you are comfortable with it, otherwise combination of relay ops is fine. Performance + support for options can be done later (by me).

masahi commented:
Looks good :+1: GPU is not supported right?

masahi commented on tests/python/relay/test_op_level3.py:
```diff
@@ -1453,5 +1453,53 @@ def verify_scatter_nd_with_stack(data_np, indices_np, shape, ref_res, rtol=1e-5,
+            backends = ["vm", "debug"]
+        else:
+            backends = ["graph", "debug"]
+        for target, ctx in tvm.testing.enabled_targets():
```
This will probably try to run on GPU

ymwangg replied:
thanks, will fix it.

masahi commented on python/tvm/topi/unique.py:
```diff
@@ -0,0 +1,118 @@
+    unique_elements = output_tensor(data.shape, data.dtype)
+    indices = output_tensor(data.shape, "int32")
+    for i in range(data.shape[0]):
+        indices[argsorted_indices[i]] = inc_scan[i]
```
We can do this loop in parallel

ymwangg replied:
actually all loops can be done in parallel, I changed all loops to parallel.

masahi commented on python/tvm/topi/unique.py:
```diff
@@ -0,0 +1,118 @@
+def _calc_adjacent_diff(data):
+    output = output_tensor(data.shape, "int32")
+    output[0] = int32(0)
+    for i in range(1, data.shape[0]):
```
Parallel

ymwangg replied:
changed

masahi commented:
Can you also add pytorch frontend? Not all option need to be supported. Likely the same as tf conversion

ymwangg commented:
@masahi Yeah, I only added CPU version in this PR. I'm not very familiar with GPU IR now but I can do it later. If the overall structure looks good, I can add `unique_with_counts` in the future since their implementations are very similar.

masahi commented:
I can do the GPU version. It will likely require ir builder. But let me know if you *want* to do GPU as well, you can certainly do it. The idea is identical with CPU version, just using different parallelization.

If `unique_with_counts` can be supported by adding another option to `unique`, that sounds good. We shouldn't add `relay.unique_with_counts` or `topi.unique_with_counts`.

ymwangg commented:
@masahi I added the `return_counts` option for the `topi.unique` operator. I also added pytorch frontend. Interestingly, it looks like pytorch returns unique elements in random order when `sorted=False`.

I'll work on the GPU version of `unique` next week.

masahi commented on python/tvm/relay/frontend/pytorch.py:
```diff
@@ -2164,6 +2164,21 @@ def is_floating_point(self, inputs, input_types):
+    def unique(self, inputs, input_types):
+        assert len(inputs) == 4
+        [data, is_sorted, return_inverse, return_counts] = inputs
+        if return_counts:
```
yeah, PyTorch doc https://pytorch.org/docs/stable/generated/torch.unique.html says their CUDA unique always returns sorted outputs regardless of `sorted` arg. So I think we can also assume that `is_sorted` is always True and add a warning if `is_sorted` arg is False.

ymwangg replied:
changed

ymwangg commented:
@masahi I added the GPU version and it's ready for review.

masahi commented on python/tvm/relay/frontend/pytorch.py:
```diff
@@ -2164,6 +2164,24 @@ def is_floating_point(self, inputs, input_types):
 
+    def unique(self, inputs, input_types):
+        assert len(inputs) == 4
+        [data, is_sorted, return_inverse, return_counts] = inputs
```
Raise an error if `return_inverse` is True, since we don't support it

ymwangg replied:
What do you mean? The implementation calculates inverse indices all the time.

masahi replied:
ah sorry you are right. So in Torchscript `unique` always returns two outputs, regardless of `return_inverse` option? The python one https://pytorch.org/docs/stable/generated/torch.unique.html can return only one output.

ymwangg replied:
Yes, I found `aten::_unique2`, the function that `torch.unique` dispatches to when `dim=None`, always return indices. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Unique.cpp#L69

masahi commented on python/tvm/relay/op/transform.py:
```diff
@@ -1463,3 +1463,48 @@ def cumsum(data, axis=None, dtype=None, exclusive=None):
+    counts (optional) : relay.Expr
+        A 1-D tensor containing the count of each unique element in the output
+    Examples
+    --------
```
Please mention that the outputs after `num_unique` are undefined.

ymwangg replied:
Added.

masahi commented on python/tvm/topi/unique.py:
```diff
@@ -0,0 +1,181 @@
+        idx = index_converter[inc_scan[i]]
+        counts[idx] += int32(1)
+    return unique_elements, indices, counts
+
```
I'm definitely seeing duplicated code here, but that's one I reason I don't like hybridscript: It doesn't have any abstraction like function

ymwangg replied:
I have changed hybrid script to ir builder similar to the GPU implementation.

masahi replied:
It looks much cleaner now, very nice. Thanks.

masahi commented on python/tvm/topi/cuda/unique.py:
```diff
@@ -0,0 +1,384 @@
+    return ib.get()
+
+
+def _calc_counts_unsorted_ir(inc_scan, index_converter, counts):
```
This looks similar to `_calc_counts_sorted_ir`, maybe you can do some trick around `index_converter` to share the implementation?

ymwangg replied:
Good suggestion, I was able to combine everything into a single ir function.

masahi commented on python/tvm/topi/cuda/unique.py:
```diff
@@ -0,0 +1,384 @@
+            index = inc_scan_ptr[tid]
+            atomic_add_return[0] = tvm.tir.call_intrin(
+                counts.dtype,
+                "tir.atomic_add",
```
hmm I haven't looked into detail at all, but atomic sounds like a code smell... I'll think about if there is a way to avoid it

ymwangg replied:
Removed to use `arange`.

masahi replied:
Looks great

masahi commented:
@ymwangg @codeislife99 I found a neat trick PyTorch uses for `count`. https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/aten/src/ATen/native/cuda/Unique.cu#L60-L68

Basically, after you get ex scan, instead of copy original inputs, you copy from an array [0, 1, 2, ....]. This will give you something like [0, 2, 5], and doing adjacent element on it directly gives the count. Does this make sense? It should be much faster than atomic.

PyTorch uses a separate `unique_by_key` call to compute counts, but since we have ex scan outputs lying around, we don't need this separate call. So we can be faster than PyTorch.

ymwangg commented:
@masahi thanks. I'll try using `arange` and `adjacent_difference` to compute the counts rather than counting by adding.

masahi commented on python/tvm/relay/frontend/tensorflow.py:
```diff
@@ -2324,6 +2324,35 @@ def _impl(inputs, attr, params, mod):
+        )
+
+    return _impl
+
```
probably we can share the implementation in two converters, just pass `return_counts` to `_unique()` and _`unique_with_counts()`

ymwangg replied:
good suggestion, changed.

masahi commented on python/tvm/topi/cuda/unique.py:
```diff
@@ -0,0 +1,394 @@
+                unique_idx = tid if not index_converter_ptr else index_converter_ptr[tid]
+                with ib.if_scope(tid == 0):
+                    counts_ptr[unique_idx] = arange_ptr[tid]
+                with ib.else_scope():
```
Probably we don't need to allocate `arange` array explicitly. Just assign `tid` should work?

ymwangg replied:
Sorry the name may be confusing. Actually the `arange_ptr` here records the cumulated counts of the unique elements and it goes like:
Step1: inc_scan = [0, 0, 1, 1, 2, 2, 2, 2, 2, 3]
                    tids = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
Step2: arange_ptr = [2, 4, 9, 10]
Step3: counts_ptr = [2, 2, 5, 1]

Maybe we can calculates `counts` first and use `indices_ptr` to replace `arange_ptr`?

masahi replied:
I didn't get why you need to fill in `arange_ptr` with cumulated counts or need a temp buffer. If you move the condition `ib.if_scope(inc_scan_ptr[tid] != inc_scan_ptr[tid - 1])` at L200 here, can't you inline the difference computation without materializing `arange_ptr`? I guess you will be writing at index `inc_scan_ptr[tid] - 1` instead of `unique_idx`.

masahi replied:
I'm imaging that we can have something like

`counts_ptr[inc_scan_ptr[tid] - 1] = tid - indices_ptr[inc_scan_ptr[tid - 1]]`

since the contents of `arange` seems redundant given other information you already have here, I have a strong feeling that we don't need to materialize `arange`.

masahi replied:
We can also consider filling in `counts` array just like you fill in `arange`, and do in-place `_calc_adjacent_diff` on it directly. 

Anyway, as long as we remove `arange` array, I'm happy.

masahi replied:
@ymwangg I made a lot of noise above, but after some thought I realized that probably we cannot remove `arange`: We cannot tell which previous index in arange would correspond to the previous unique element, without materializing `arange` array and doing global sync. In-place adjacent diff obvious doesn't work without some synchronization. 

So maybe we rename `arange` to something like `unique_seq_indices` and keep everything else as is, that would be good?

ymwangg replied:
Right, the main problem is for a `tid` with condition `inc_scan[tid] != inc_scan [tid]` we also need to know the previous `tid` with the same condition. It's possible to use `count` to store the cumulative sum but getting the adjacent diff in-place requires global sync.

I think we can use `indices_ptr` as a tmp buf to avoid allocating a new array.

masahi replied:
I'd hope that TVM could automatically do such buffer reuse optimization, but not sure if it would work in this case (can check by dumping the cuda source). I'm also ok with having `arange` explicitly for clarity.

ymwangg replied:
I checked the cuda kernels and if they are executed in the order they appear in IR (I think they do), then I don't see any potential issues.

masahi commented on python/tvm/relay/frontend/tensorflow.py:
```diff
@@ -2324,6 +2324,39 @@ def _impl(inputs, attr, params, mod):
+
+
+def _unique_with_counts():
+    def _impl(inputs, attr, params, mod):
```
Remove this

ymwangg replied:
removed, thanks for the catch.

masahi reviewed (approved):
LGTM :+1:

masahi commented:
Thanks @ymwangg @codeislife99, this is really a great work!

ymwangg commented:
@masahi thanks for making this such an interesting project!
